pkg/controller/pravegacluster/upgrade.go

```diff
@@ -339,31 +415,25 @@ func (r *ReconcilePravegaCluster) syncBookkeeperVersion(p *pravegav1alpha1.Prave
 			return false, err
 		}
 
-		// Set the upgrade condition reason to be UpgradingBookkeeperReason, message to be 0
-		p.Status.SetUpgradingConditionTrue(pravegav1alpha1.UpgradingBookkeeperReason, "0")
-
 		// Updated pod template
 		return false, nil
 	}
 
 	// Pod template already updated
-
 	log.Printf("statefulset (%s) status: %d updated, %d ready, %d target", sts.Name,
 		sts.Status.UpdatedReplicas, sts.Status.ReadyReplicas, sts.Status.Replicas)
 
 	// Check whether the upgrade is in progress or has completed
 	if sts.Status.UpdatedReplicas == sts.Status.Replicas &&
 		sts.Status.UpdatedReplicas == sts.Status.ReadyReplicas {
 		// StatefulSet upgrade completed
-		// TODO: wait until there is no under replicated ledger
```

Comment:
Because there us nothing to be done for under-replicated ledgers here. Autorecovery takes care of that.

Reply:
Auto-replication in BK does take care of replicating data, but we don't actually know when that happens for any given ledger. Consequently, a ledger can be under-replicated for arbitrarily long.

I believe this comment is saying that, before declaring an upgrade complete, we should make sure that all under-replicated ledgers have been properly replicated.

Reply:
> Auto-replication in BK does take care of replicating data, but we don't actually know when that happens for any given ledger. Consequently, a ledger can be under-replicated for arbitrarily long.
> 
> I believe this comment is saying that, before declaring an upgrade complete, we should make sure that all under-replicated ledgers have been properly replicated.

Auto replication kicks in if a bookie is down for more than a certain configurable time interval, and makes sure under-replicated ledgers are dealt with. I have documented the details of this in #235.   
We would need to tune auto-recovery configuration parameters based on how long we want to tolerate data loss or how soon we need the recovery to start.
Nevertheless, this not an upgrade problem but a general DR issue, where a certain server or set of servers could be down for an arbitrary period of time (due to upgrade or otherwise) and data from the same needs to be recovered.

Reply:
>  this not an upgrade problem but a general DR issue

there is a difference between data being unavailable or under-replicated due to say a crash and the data being unavailable or under-replicated due to a planned task like an upgrade. maybe this is not for this PR, but I don't think it is acceptable that post upgrade the cluster is in a worse state (e.g., under-replicated data) compared to what it was prior to upgrade starting.

Reply:
Note that under-replication introduced by upgrade is temporary (till the node being upgraded comes back - about 2-3 mins ) and should be acceptable.
Bookkeeper documentation infact suggests switching off Autorecovery during maintenance tasks like upgrade, as the under-replication is only temporary and should not cause auto-recovery to trigger unnecessarily:
https://bookkeeper.apache.org/docs/4.7.3/admin/autorecovery/
```
You can disable AutoRecovery at any time, for example during maintenance. Disabling AutoRecovery ensures that bookies’ data isn’t unnecessarily rereplicated when the bookie is only taken down for a short period of time, for example when the bookie is being updated or the configuration if being changed.```

Reply:
I don't see how a cluster could be in a worse state post upgrade. 
Data replication is really not related to upgrade.
The right auto-recovery configuration would make sure under-replicated data is replicated every 'x' mins, irrespective of whether a node is down due to upgrade or other issues.

Reply:
In terms of limiting the damage upgrade can cause due to nodes being down, I have capped the max time a node can take to come back up to 10 mins, failing which we fail the entire upgrade process and rollback should bring the cluster back to a normal state.

Reply:
I keep forgetting that we don't wipe out the bookie data. Perhaps we should consider turning the auto-replication off during an upgrade as you mention, @pbelgundi . We should create an issue so that we can assess whether it is necessary.

Reply:
Issue for tracking auto-recovery enable/disable:
https://github.com/pravega/pravega-operator/issues/267